packages/protobuf/src/next/to-binary.ts

```diff
@@ -0,0 +1,236 @@
+import type { Message } from "./types.js";
+import type { BinaryWriteOptions } from "../binary-format.js";
+import { reflect } from "./reflect/reflect.js";
+import {
+  BinaryWriter,
+  WireType,
+  type IBinaryWriter,
+} from "../binary-encoding.js";
+import {
+  FeatureSet_FieldPresence,
+  FeatureSet_MessageEncoding,
+  FieldDescriptorProto_Label,
+  FieldDescriptorProto_Type,
+} from "../google/protobuf/descriptor_pb.js";
+import { ScalarType, type ScalarValue } from "./reflect/scalar.js";
+import type { DescField } from "../descriptor-set.js";
+
+// Default options for serializing binary data.
+const writeDefaults: Readonly<BinaryWriteOptions> = {
+  writeUnknownFields: true,
+  writerFactory: () => new BinaryWriter(),
+};
+
+function makeWriteOptions(
+  options?: Partial<BinaryWriteOptions>,
+): Readonly<BinaryWriteOptions> {
+  return options ? { ...writeDefaults, ...options } : writeDefaults;
+}
+
+export function toBinary<T extends Message>(
+  message: T,
+  options?: Partial<BinaryWriteOptions>,
+): Uint8Array {
+  return reflectToBinary(reflect(message), makeWriteOptions(options));
+}
+
+function reflectToBinary(
+  msg: ReflectMessage,
+  opts: BinaryWriteOptions,
+): Uint8Array {
+  const w = opts.writerFactory();
+  for (const f of msg.fields) {
+    if (!msg.isSet(f)) {
+      if (
+        f.proto.label === FieldDescriptorProto_Label.REQUIRED ||
+        f.getFeatures().fieldPresence ==
+          FeatureSet_FieldPresence.LEGACY_REQUIRED
+      ) {
+        throw new Error(
+          `cannot encode field ${msg.desc.typeName}.${f.name} to binary: required field not set`,
+        );
+      }
+      continue;
+    }
+    switch (f.fieldKind) {
+      case "scalar":
+      case "enum":
+        writeScalar(w, f.scalar ?? ScalarType.INT32, f.number, msg.get(f));
+        break;
+      case "list":
+        writeListField(w, opts, f, msg.get(f));
+        break;
+      case "message":
+        writeMessageField(w, opts, f, msg.get(f) as Message);
+        break;
+      case "map":
+        for (const [key, val] of Object.entries(msg.get(f))) {
+          writeMapEntry(w, opts, f, key, val);
+        }
+        break;
+    }
+  }
+  return w.finish();
+}
+
+function writeScalar(
+  writer: IBinaryWriter,
+  scalarType: ScalarType,
+  fieldNo: number,
+  value: unknown,
+) {
+  const [wireType, method] = scalarTypeInfo(scalarType);
+  (writer.tag(fieldNo, wireType)[method] as (v: unknown) => void)(value);
+}
+
+function writeMessageField(
+  writer: IBinaryWriter,
+  opts: BinaryWriteOptions,
+  field: DescField &
+    ({ fieldKind: "message" } | { fieldKind: "list"; listKind: "message" }),
+  message: Message,
+) {
+  const rm = reflectToBinary(reflect(message), opts);
+  if (
+    field.proto.type === FieldDescriptorProto_Type.GROUP ||
+    field.getFeatures().messageEncoding === FeatureSet_MessageEncoding.DELIMITED
+  ) {
+    writer
+      .tag(field.number, WireType.StartGroup)
+      .raw(rm)
```

Comment:
We're creating a new writer here for a a message field that uses the delimited encoding. The advantage of the delimited encoding is that we don't need to do that, because we don't need to write a length prefix. 

We can optimize this once we have more test coverage. Can you add a TODO?